### PR TITLE
[Search Pipeline]Handled default search pipeline for multiple indices

### PR DESCRIPTION
**Pull request by owaiskazi19**
### Description
In https://github.com/opensearch-project/OpenSearch/pull/7470, we added the ability to specify a search pipeline that will execute by default for queries against a single index. 
This PR handles a check to see if all the concrete indices have the same default pipeline and attached it.   

### Related Issues
Resolves https://github.com/opensearch-project/OpenSearch/issues/7512

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Failing checks are inspected and point to the corresponding known issue(s) (See: [Troubleshooting Failing Builds](../blob/main/CONTRIBUTING.md#troubleshooting-failing-builds))
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))
- [x] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by owaiskazi19** [comment]
> ❌ Gradle check result for [08fa4d3](https://github.com/opensearch-project/OpenSearch/commit/08fa4d3e28bd8951a95fbc6a2d68b7f21750a03a): [FAILURE](https://build.ci.opensearch.org/job/gradle-check/37262/)
> 
> Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green. Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?

https://github.com/opensearch-project/OpenSearch/issues/13278

File: server/src/main/java/org/opensearch/search/pipeline/SearchPipelineService.java [hunk 1]
```diff
@@ -360,7 +362,7 @@ static ClusterState innerDelete(DeleteSearchPipelineRequest request, ClusterStat
         return newState.build();
     }
 
-    public PipelinedRequest resolvePipeline(SearchRequest searchRequest) {
+    public PipelinedRequest resolvePipeline(SearchRequest searchRequest, IndexNameExpressionResolver indexNameExpressionResolver) {
```
Comment (dbwiddis) [inline server/src/main/java/org/opensearch/search/pipeline/SearchPipelineService.java:365]: We are changing a public method here without providing a backward compatible overload. It's not clear whether this file is API or internal.  I did do a brief search of all our repos and don't see it called anywhere else, but we should probably either make this compatible with an overload or add the opensearch-internal annotation to this class.

File: server/src/main/java/org/opensearch/search/pipeline/SearchPipelineService.java [hunk 2]
```diff
@@ -390,13 +392,22 @@ public PipelinedRequest resolvePipeline(SearchRequest searchRequest) {
             if (searchRequest.pipeline() != null) {
                 // Named pipeline specified for the request
                 pipelineId = searchRequest.pipeline();
-            } else if (state != null && searchRequest.indices() != null && searchRequest.indices().length == 1) {
+            } else if (state != null && searchRequest.indices() != null) {
```
Comment (dbwiddis) [inline server/src/main/java/org/opensearch/search/pipeline/SearchPipelineService.java:395]: We need to keep a non-zero check here, see next comment.

File: server/src/main/java/org/opensearch/search/pipeline/SearchPipelineService.java [hunk 3]
```diff
@@ -390,13 +392,22 @@ public PipelinedRequest resolvePipeline(SearchRequest searchRequest) {
...
                 // Check for index default pipeline
-                IndexMetadata indexMetadata = state.metadata().index(searchRequest.indices()[0]);
-                if (indexMetadata != null) {
-                    Settings indexSettings = indexMetadata.getSettings();
-                    if (IndexSettings.DEFAULT_SEARCH_PIPELINE.exists(indexSettings)) {
-                        pipelineId = IndexSettings.DEFAULT_SEARCH_PIPELINE.get(indexSettings);
+                Index[] concreteIndices = indexNameExpressionResolver.concreteIndices(state, searchRequest);
```
Comment (dbwiddis) [inline server/src/main/java/org/opensearch/search/pipeline/SearchPipelineService.java:397]: If `searchRequest.indices()` is empty here, it goes to this line of code and applies to _all_ indices:
```java
    Index[] concreteIndices(Context context, String... indexExpressions) {
        if (indexExpressions == null || indexExpressions.length == 0) {
            indexExpressions = new String[] { Metadata.ALL };
        }
```
I don't think that's what we want.

And we should have a unit test on this edge case.

File: server/src/test/java/org/opensearch/search/pipeline/SearchPipelineServiceTests.java [hunk 4]
```diff
@@ -96,14 +99,31 @@ public Map<String, Processor.Factory<SearchPhaseResultsProcessor>> getSearchPhas
...
 
+    private static IndexMetadata.Builder indexBuilder(String index) {
+        return indexBuilder(index, Settings.EMPTY);
+    }
+
+    private static IndexMetadata.Builder indexBuilder(String index, Settings additionalSettings) {
+        return IndexMetadata.builder(index).settings(settings(additionalSettings));
+    }
+
+    private static Settings.Builder settings(Settings additionalSettings) {
```
Comment (dbwiddis) [inline server/src/test/java/org/opensearch/search/pipeline/SearchPipelineServiceTests.java:121]: This seems to do a bit more internally with the settings than the method name/argument implies.

**Review by dbwiddis — DISMISSED** [review]

File: server/src/main/java/org/opensearch/search/pipeline/SearchPipelineService.java (hunk 1 above)
Comment (owaiskazi19) [reply server/src/main/java/org/opensearch/search/pipeline/SearchPipelineService.java:365]: @froh wdyt of adding `@opensearch.internal` to this class?

File: server/src/test/java/org/opensearch/search/pipeline/SearchPipelineServiceTests.java (hunk 4 above)
Comment (owaiskazi19) [reply server/src/test/java/org/opensearch/search/pipeline/SearchPipelineServiceTests.java]: Changed it to `addAdditionalSettings`

File: server/src/main/java/org/opensearch/search/pipeline/SearchPipelineService.java (hunk 3 above)
Comment (owaiskazi19) [reply server/src/main/java/org/opensearch/search/pipeline/SearchPipelineService.java:397]: Added the check and a unit test

File: server/src/main/java/org/opensearch/search/pipeline/SearchPipelineService.java [hunk 5]
```diff
@@ -390,13 +392,22 @@ public PipelinedRequest resolvePipeline(SearchRequest searchRequest) {
...
+                for (Index index : concreteIndices) {
+                    IndexMetadata indexMetadata = state.metadata().index(index);
+                    if (indexMetadata != null) {
+                        Settings indexSettings = indexMetadata.getSettings();
+                        if (IndexSettings.DEFAULT_SEARCH_PIPELINE.exists(indexSettings)) {
+                            String currentPipelineId = IndexSettings.DEFAULT_SEARCH_PIPELINE.get(indexSettings);
+                            if (NOOP_PIPELINE_ID.equals(pipelineId)) {
+                                pipelineId = currentPipelineId;
+                            } else if (pipelineId.equals(currentPipelineId) == false) {
```
Comment (dbwiddis) [inline server/src/main/java/org/opensearch/search/pipeline/SearchPipelineService.java:406]:
```suggestion
                            } else if (!pipelineId.equals(currentPipelineId)) {
```

File: server/src/main/java/org/opensearch/search/pipeline/SearchPipelineService.java (hunk 3 above)
Comment (dbwiddis) [reply server/src/main/java/org/opensearch/search/pipeline/SearchPipelineService.java:397]: As you've seen in your jenkins test failure, this is throwing an exception if an index name doesn't resolve, so this whole section needs to be in a try/catch block (ignoring the unmatched index and just not applying a default pipeline)

**Review by dbwiddis — COMMENTED** [review]
LGTM pending fix for index not found exception

File: server/src/main/java/org/opensearch/search/pipeline/SearchPipelineService.java (hunk 3 above)
Comment (owaiskazi19) [reply server/src/main/java/org/opensearch/search/pipeline/SearchPipelineService.java]: Yes, added a try/catch block to handle the exception. Also, a added a UT for invalid index

**Comment by owaiskazi19** [comment]
> ❌ Gradle check result for [a57abae](https://github.com/opensearch-project/OpenSearch/commit/a57abaecca61cd790e602b33b8431a353ab19cd7): [FAILURE](https://build.ci.opensearch.org/job/gradle-check/37334/)
> 
> Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green. Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?

https://github.com/opensearch-project/OpenSearch/issues/12639

File: server/src/main/java/org/opensearch/search/pipeline/SearchPipelineService.java (hunk 1 above)
Comment (msfroh) [reply server/src/main/java/org/opensearch/search/pipeline/SearchPipelineService.java:365]: Yes -- we should label it `@opensearch.internal`. I didn't know enough about the annotations when I added it.

**Comment by owaiskazi19** [comment]
> ❌ Gradle check result for [ad6ff2f](https://github.com/opensearch-project/OpenSearch/commit/ad6ff2f5856c051ce2a1875d778b7aa562f00937): [FAILURE](https://build.ci.opensearch.org/job/gradle-check/37386/)
> 
> Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green. Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?

https://github.com/opensearch-project/OpenSearch/issues/13240

**Comment by owaiskazi19** [comment]
@dbwiddis @msfroh can I get another review please? Thanks

File: server/src/main/java/org/opensearch/search/pipeline/SearchPipelineService.java [hunk 6]
```diff
@@ -390,14 +394,27 @@ public PipelinedRequest resolvePipeline(SearchRequest searchRequest) {
...
+                                    pipelineId = currentPipelineId;
+                                } else if (!pipelineId.equals(currentPipelineId)) {
+                                    pipelineId = NOOP_PIPELINE_ID;
+                                    break;
+                                }
+                            }
+                        }
                     }
+                } catch (Exception e) {
+                    logger.info("Default pipeline not applied for {}", (Object) searchRequest.indices());
```
Comment (dbwiddis) [inline server/src/main/java/org/opensearch/search/pipeline/SearchPipelineService.java:417]: The described log entry seems like an extremely common occurrence (search an index without a default pipeline). 

This error is actually only from the index name expression resolver, right?  In which case can we narrow the exception type and use the `searchRequest` to extract the requested indices.

Info level logging is possibly too high a level, I'd suggest debug.

**Review by dbwiddis — APPROVED** [review]
LGTM!

File: server/src/main/java/org/opensearch/search/pipeline/SearchPipelineService.java [hunk 7]
```diff
@@ -390,14 +395,27 @@ public PipelinedRequest resolvePipeline(SearchRequest searchRequest) {
...
+                                } else if (!pipelineId.equals(currentPipelineId)) {
+                                    pipelineId = NOOP_PIPELINE_ID;
+                                    break;
+                                }
+                            }
+                        }
                     }
+                } catch (IndexNotFoundException e) {
+                    logger.debug("Default pipeline not applied for {}", (Object) searchRequest.indices());
                 }
```
Comment (msfroh) [inline server/src/main/java/org/opensearch/search/pipeline/SearchPipelineService.java:419]: I don't think we can just swallow the exception without changing any state.

In particular, suppose the request targets indices A, B, and C, where A and C have default (but different) pipelines and B doesn't exist. We would end up applying A's pipeline to C (assuming `ignore_unavailable` is `true`, so we skip B).

Or is the exception thrown by `concreteIndices`? If that's the case we might do the wrong thing when `ignore_unavailable` is `true` -- in the example above, if A and C use the same pipeline and B doesn't exist, I think we *would* want to apply the pipeline (since we're just ignoring B).

Comment (owaiskazi19) [reply server/src/main/java/org/opensearch/search/pipeline/SearchPipelineService.java:419]: Yes the exception is thrown from `concreteIndices` and it fails the below test if we don't have a try/catch here

```
REPRODUCE WITH: ./gradlew ':qa:ccs-unavailable-clusters:integTest' --tests "org.opensearch.search.CrossClusterSearchUnavailableClusterIT.testSearchSkipUnavailable" -Dtests.seed=D240A247ABFCD6B7 -Dtests.security.manager=true -Dtests.jvm.argline="-XX:TieredStopAtLevel=1 -XX:ReservedCodeCacheSize=64m" -Dtests.locale=de-LU -Dtests.timezone=Pacific/Wake -Druntime.java=17
  2> [remote1:index] OpenSearchStatusException[OpenSearch exception [type=index_not_found_exception, reason=no such index [remote1:index]]]
        at __randomizedtesting.SeedInfo.seed([D240A247ABFCD6B7:8532D8AF58B6D5E]:0)
        at app//org.opensearch.rest.BytesRestResponse.errorFromXContent(BytesRestResponse.java:210)
        at app//org.opensearch.client.RestHighLevelClient.parseEntity(RestHighLevelClient.java:2235)
        at app//org.opensearch.client.RestHighLevelClient.parseResponseException(RestHighLevelClient.java:2212)
        at app//org.opensearch.client.RestHighLevelClient.internalPerformRequest(RestHighLevelClient.java:1931)
        at app//org.opensearch.client.RestHighLevelClient.performRequest(RestHighLevelClient.java:1884)
        at app//org.opensearch.client.RestHighLevelClient.performRequestAndParseEntity(RestHighLevelClient.java:1852)
        at app//org.opensearch.client.RestHighLevelClient.search(RestHighLevelClient.java:1095)
        at app//org.opensearch.search.CrossClusterSearchUnavailableClusterIT.testSearchSkipUnavailable(CrossClusterSearchUnavailableClusterIT.java:177)
  2> NOTE: leaving temporary files on disk at: /home/ubuntu/OpenSearch/qa/ccs-unavailable-clusters/build/testrun/integTest/temp/org.opensearch.search.CrossClusterSearchUnavailableClusterIT_D240A247ABFCD6B7-001
```

Comment (msfroh) [reply server/src/main/java/org/opensearch/search/pipeline/SearchPipelineService.java:419]: I think the fundamental problem here was that I misunderstood the behavior of `concreteIndices`.

Since it takes the `SearchRequest` as input and the `SearchRequest` has the `indicesOptions` property, then `concreteIndices` will not throw an exception if `ignore_unavailable` is `true` -- it will just silently skip B without throwing an exception. So, the behavior I described above should happen "for free".

Part of me is still a little nervous about swallowing the exception, but it will still get thrown later, when we try to determine what shards to target. So, it should be good. :+1:

**Review by msfroh — APPROVED** [review]

**Comment by owaiskazi19** [comment]
> ❌ Gradle check result for [e393181](https://github.com/opensearch-project/OpenSearch/commit/e3931818ba3fe10771e7b61695526c431f15b05a): [FAILURE](https://build.ci.opensearch.org/job/gradle-check/37892/)
> 
> Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green. Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?

https://github.com/opensearch-project/OpenSearch/issues/12197

**Comment by dbwiddis** [comment]
Failed gradle check tracked in https://github.com/opensearch-project/OpenSearch/issues/13437

**Review by andrross — APPROVED** [review]
